blog/_posts/2016-12-xx-moredots.md

```diff
@@ -0,0 +1,449 @@
+
+* What is the overhead of traditional "vectorized" code?  Isn't vectorized code supposed to be fast already?
+
+* Why are all these dots necessary?  Couldn't Julia just optimize "ordinary" vector code?
+
+* Is this something unique to Julia, or can other languages do the same thing?
+
+The short answers are:
+
+* Ordinary vectorized code is fast, but not as fast as a hand-written loop
+  (assuming loops are efficiently compiled)
+  because each vectorized operation generates a new temporary array and
+  executes a separate loop, leading to a lot of overhead when multiple
+  vectorized operations are combined.
+
+* The dots allow Julia to recognize the "vectorized" nature of the
+  operations at a *syntactic* level (before e.g. the type of `x` is known),
+  and hence the loop fusion is a *syntactic guarantee*, not a
+  compiler optimization that may or may not occur for carefully written code.
+
+* Other languages have implemented loop fusion for vectorized operations,
+  but typically for only a small set of types and operations/functions that
+  are known to the compiler.  Julia's ability to do it generically, even
+  for *user-defined* array types and functions/operators, is unusual
+  and relies in part on the syntax choices above and on its ability to efficiently
+  compile higher-order functions.
+
+## Isn't vectorized code already fast?
+
+Let's rewrite the code above in a more traditional vectorized style, without
+so many dots, that you might use in Julia 0.4 or in other languages
+(most famously Matlab, Python/Numpy, or R).   
+
+```jl
+X = f(2 * X.^2 + 6 * X.^3 - sqrt(X))
+```
+
+Of course, this assumes that the functions `sqrt` and `f` are "vectorized,"
+i.e. that they take vector arguments `X` and compute the
+function elementwise.  This is true of `sqrt` in Julia 0.4, but it
+means that we have to rewrite our own function `f` in a vectorized style, as
+e.g. `f(x) = 3x.^2 + 5x + 2` (using the elementwise operator `.^` because
+`array^scalar` is not defined).   (If we cared a lot about efficiency, we
+might instead define a special method `f(X::AbstractArray) = map(f, X)`
+or use the `@vectorize_1arg f Number` macro defined in Julia 0.4.)
+
+### Which functions are vectorized?
+
+As an aside, this example illustrates an annoyance with the vectorized style:
+
+* You have to decide *in advance* whether a given function `f(x::Number)`
+  will also be applied elementwise to arrays, and define a corresponding elementwise method.
+
+For library functions like `sqrt`, this means that the library authors
+have to guess at which functions should have vectorized methods, and users
+have to guess at what vaguely defined subset of library functions work
+for vectors.
+
+One possible solution is to vectorize *every function automatically*.   The
+language [Chapel](https://en.wikipedia.org/wiki/Chapel_%28programming_language%29)
```

Comment:
Not the same thing, I think.  In Matlab, scalars are 1-element vectors too, but this is not the same thing as what Chapel does either.   In Julia, you can declare a function as `f(x) = ...` and it will accept any argument type, but it won't necessarily *work* for any argument type.

The problem is that the same code doesn't generally work for scalars as for vectors.   e.g. in the example in the blog post, `f(x) = 3x^2 + 5x + 2`doesn't work for vectors because `x^2` isn't defined for vectors.   In R, if you have a function like `if (x < 0) { 2*x } else { 3*x }` then it will do *something* for vectors, but probably not what you want (it will only look at the first element of `x < 0`).

Chapel, in contrast, effectively defines a generic function `(f::Function)(args::AbstractArray...) = map(f, args...)` that is called as a fallback, so if you just define `f(x::Number) = x < 0 ? 2*x : 3*x` it will do the elementwise thing for `f(::Array)`.